tests/blackbox.py

```diff
@@ -192,7 +196,7 @@ def handle_arg(arg):
 
     def abi_return_type(self) -> None | algosdk.abi.ABIType:
         if not self.is_abi():
-            raise NotImplementedError("this {type(self)} is not A.B.I. compatible")
```

Comment:
@tzaffi Highlighting a change I'm unsure about.

For ease of use, the PR relies on `abi_return_type` (and `abi_argument_types`) to return `None` when the caller is _not_ requesting an ABI dryrun.

In my inspection, I saw _no_ reason the exception need be thrown here.  So, I changed the behavior.  Though I'm less familiar with the logic than you, and I request you double check the change.

Reply:
Though the previous is admittedly clunky, I don't support this change. The reason is that the `abi_*` methods only make sense on `PyTealDryRunExecutor`'s which have been instantiated on `@ABIReturnSubroutine` decorated methods (as opposed to regular `@Subroutine`'s). Perhaps this calls for a class-hierarchy whereby the `@ABIReturnSubroutine`-compliant objects subclass the more generic ones, but it's probably overkill in this testing context. But I'm open to further discussion.

Reply:
@tzaffi I've attempted to add requested unit tests in https://github.com/algorand/pyteal/pull/338/commits/cf938ae962e4499c294f0b39d1e5efcd655b6fd3.  If it looks good to you, can you resolve the thread?